fix(sqlserver): queries returning multiple record sets

lboonekamp opened the pull request:
### Description of change

Fixes #11566, fixes #5051

#### What?
The change resolves the issue whereby any call to `query()` for mssql connections, returns only the first recordset, even when a query returns multiple recordsets. This change checks for recordsets and adds it to the `QueryResult` if `useStructuredQuery = true` is passed in. This change will allow the passing in of a boolean, anywhere `query()` is called, which will enable the full list of results to be accessed. This change does not change the default behaviour.

#### Why?
 Without this change, you cannot access all recordsets in the return, you only get the first one.

#### Verification
Run a stored procedure that returns multiple recordsets i.e. `.query('MY_SP')`.
Run two select statements within a single call to `.query('SELECT 1; SELECT 2')`.

### Pull-Request Checklist

-   [x] Code is up-to-date with the `master` branch
-   [x] This pull request links relevant issues as `Fixes #11566, #5051`
-   [x] There are new or updated unit tests validating the change
-   [x] Documentation has been updated to reflect this change


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Introduced QueryOptions to request structured results from raw queries across data sources, entity managers, and query runners.
  - getRawMany now accepts options to return structured results when requested.
  - Query results now expose recordsets for multi-statement responses.

- Deprecations
  - The boolean overload for requesting structured results is deprecated; use QueryOptions instead.

- Documentation
  - Added guidance and examples for executing raw SQL with structured results, including migration notes and API usage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lboonekamp committed:
Added history to .gitignore

lboonekamp committed:
feat: Add useStructuredResult and recordsets

-Add useStructuredResult to EntityManager.query() and QueryRunner.query()

lboonekamp committed:
fixes #11566, fixes #5051

lboonekamp committed:
Updating docs

lboonekamp committed:
Fixing broken test for manager.query()

lboonekamp committed:
Removing commented out blocks

lboonekamp committed:
docs: Removed queryRunner argument

from EntityManager docs

lboonekamp committed:
Removed duplicate test

alumni commented on .gitignore:
```diff
@@ -29,3 +29,5 @@ ormlogs.log
 # test projects generated by test/github-issues/issue-8975.ts
 /*TestProject
+
+.history
```
unnecessary change?

alumni commented on src/data-source/DataSource.ts:
```diff
@@ -538,7 +539,11 @@ export class DataSource {
 
         try {
-            return await usedQueryRunner.query(query, parameters) // await is needed here because we are using finally
+            return (await usedQueryRunner.query(
+                query,
+                parameters,
+                (useStructuredResult || false) as any,
+            )) as T // await is needed here because we are using finally
```
In addition to that: might as well add the generic to `QueryRunner.query`

lboonekamp commented:
> I wonder whether we should make the last parameter of `query` an object of type `QueryOptions` containing the property `useStructuredResult?: boolean`.
> 
> Currently, in `QueryRunner.query` the last parameter is boolean just for this option, but the database clients that TypeORM uses can accept options for executing statements (e.g. to return rows are arrays instead of objects).

Great idea. Objects are much more extendable. Let me know if you want me to make the changes.

sgarner commented on src/data-source/DataSource.ts:
```diff
@@ -528,6 +528,7 @@ export class DataSource {
         query: string,
         parameters?: any[],
         queryRunner?: QueryRunner,
+        useStructuredResult?: boolean,
```
I agree with @alumni that an options object would be preferable to a boolean.

It's also critical that the return type correctly reflects this option, which you can do using a method override as suggested by coderabbit.

sgarner commented on docs/docs/working-with-entity-manager/5-entity-manager-api.md:
```diff
@@ -62,6 +62,7 @@ const rawData = await manager.query(
 const rawData = await manager.query(
     "SELECT * FROM USERS WHERE name = @0 and age = @1",
     ["John", 24],
+    true, // optional flag to return a structured result set object containing all recordsets
```
I don't think this is the right place to show this flag, we don't want to suggest that it should be used by default when querying mssql.

And the structured result feature also applies to other drivers, even if they don't all support multiple recordsets.

Suggest we add another paragraph describing the `query` options object just below this example.

sgarner commented on docs/docs/advanced-topics/2-transactions.md:
```diff
@@ -74,6 +74,18 @@ await queryRunner.connect()
 // now we can execute any queries on a query runner, for example:
 await queryRunner.query("SELECT * FROM users")
 
+// **NOTE:** if you need to access multiple recordsets from a single query, pass true as the third parameter to the `query` method:
```
The topic of this doc page is Transactions. This is not the right place to document the structured result feature.

There is a [Query Runner](https://typeorm.io/docs/query-runner) page where it could go instead.

sgarner reviewed (changes requested):
A few improvements can be made but otherwise this looks good, thanks @lboonekamp !

Should we add this option to `SelectQueryBuilder.getRawMany()` as well?

gioboa committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

lboonekamp committed:
1. Changed useStructuredResult to be passed within a QueryOptions object.
2. Made the signature for query support 3 or 4 arguments.
3. Moved documentation to better locations.
4. Done a full migration of all the builder calls that were passing a boolean for useStructuredResult, to the new spec.
5. Improved typing for all my changes.
6. Moved tests to the query-runner\structured-result.ts file.

lboonekamp committed:
Merge branch '11566-sql-recordset-return-fix' of https://github.com/lboonekamp/typeorm into 11566-sql-recordset-return-fix

lboonekamp committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

lboonekamp committed:
Somehow half my changes were committed. I've fixed these up now.

lboonekamp committed:
Fixing incorrect assertio

lboonekamp committed:
Fixing spacing in docs

lboonekamp committed:
Fixing spacing in select qb docs

lboonekamp committed:
Added recordsets to QueryResult. I've double-checked all the required changes again as it seems none of what I changed was actually committed.

lboonekamp committed:
Rename test

lboonekamp commented:
Super annoying. Almost like I didn't commit properly. I'll push an update shortly to fix this up.

lboonekamp committed:
Removed files from commit

lboonekamp committed:
Merge branch '11566-sql-recordset-return-fix' of https://github.com/lboonekamp/typeorm into 11566-sql-recordset-return-fix

lboonekamp committed:
Merge branch '11566-sql-recordset-return-fix' of https://github.com/lboonekamp/typeorm into 11566-sql-recordset-return-fix

lboonekamp committed:
Simplified test stub and removed conditional

lboonekamp committed:
Removed duplicate test.

lboonekamp committed:
Removed test for entityManager.query() due to delegation to dataSource.query()

lboonekamp committed:
Restoring query-runner docs. Accidental deletion.

lboonekamp committed:
Restored eslint to original version and removed .history from gitignore.

lboonekamp committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

lboonekamp committed:
Fixed redundant ternary operator. Wasn't correctly returning QueryResult.

BerenLongstreet commented:
Yeah that code rabbit suggestion doesn't make any sense

lboonekamp committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

lboonekamp committed:
Merge branch '11566-sql-recordset-return-fix' of https://github.com/lboonekamp/typeorm into 11566-sql-recordset-return-fix

lboonekamp committed:
Fixed package-log. Contained an ADO repo entry. Only a single entry, which is odd.

lboonekamp committed:
Removing unused import.

lboonekamp committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

gioboa reviewed (commented):
Can I ask you to solve the conflicts please?

lboonekamp committed:
Merge upstream/master and resolve conflicts

lboonekamp committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

lboonekamp committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

alumni commented:
The tests need to end in `.test.ts`.

Also: Since this is going to make it into 1.0, breaking changes are accepted (they should be documented in migration to v1 doc page though)

lboonekamp committed:
Renamed structured result test.

lboonekamp committed:
Added migration info for query() method.

lboonekamp committed:
Merge branch '11566-sql-recordset-return-fix' of https://github.com/lboonekamp/typeorm into 11566-sql-recordset-return-fix

gioboa committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

gioboa committed:
Merge branch 'master' into 11566-sql-recordset-return-fix

gioboa reviewed (changes requested):
Tests are failing, can you fix the issues please?

pkuczynski committed:
Merge remote-tracking branch 'origin/master' into pr-11576

# Conflicts:
#	docs/docs/guides/8-migration-v1.md
#	docs/docs/working-with-entity-manager/5-entity-manager-api.md
#	src/driver/aurora-mysql/AuroraMysqlQueryRunner.ts
#	src/driver/aurora-postgres/AuroraPostgresQueryRunner.ts
#	src/driver/cockroachdb/CockroachQueryRunner.ts
#	src/driver/mysql/MysqlQueryRunner.ts
#	src/driver/oracle/OracleQueryRunner.ts
#	src/driver/postgres/PostgresQueryRunner.ts
#	src/driver/sap/SapQueryRunner.ts
#	src/driver/spanner/SpannerQueryRunner.ts
#	src/driver/sqlite/SqliteQueryRunner.ts
#	src/driver/sqlserver/SqlServerQueryRunner.ts
#	src/index.ts
#	src/query-builder/SelectQueryBuilder.ts
#	src/query-runner/QueryRunner.ts

pkuczynski committed:
Merge remote-tracking branch 'origin/master' into pr-11576

pkuczynski committed:
fix(sqlserver): revert unnecessary .gitignore change

pkuczynski committed:
fix(sqlserver): move structured-result test to unit tests

pkuczynski committed:
fix(sqlserver): use before/after pattern in migration guide

pkuczynski committed:
fix(sqlserver): remove deprecation notes from docs (v1 is a major release)

Commit:
Merge remote-tracking branch 'origin/master' into 11566-sql-recordset-return-fix

# Conflicts:
#	docs/docs/guides/8-migration-v1.md
#	src/data-source/DataSource.ts
#	src/driver/better-sqlite3/BetterSqlite3QueryRunner.ts
#	src/driver/capacitor/CapacitorQueryRunner.ts
#	src/driver/mysql/MysqlQueryRunner.ts
#	src/driver/postgres/PostgresQueryRunner.ts
#	src/entity-manager/EntityManager.ts
